packages/react-server-dom-vite/src/ClientProxy.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import {createElement} from 'react';
+
+declare var globalThis: any;
+// eslint-disable-next-line no-unused-vars
+/*global globalThis*/
+
+// This is a store of components discovered during RSC
+// to load them later when consuming the response in SSR.
+globalThis.__COMPONENT_INDEX = {};
+
+type ClientProxy = {
+  id: string,
+  name: string,
+  named: boolean,
+  component: mixed,
+};
+
+function isReactComponent(component: any, name: string) {
+  if (!component) return false;
+
+  return (
+    (typeof component === 'function' && /^[A-Z]/.test(name)) ||
+    typeof component.render === 'function' ||
+    component.$$typeof === Symbol.for('react.element')
+  );
+}
+
+// A ClientProxy behaves as a module reference for the Flight
+// runtime (RSC) and as a real component for the Fizz runtime (SSR).
+// Note that this is not used in browser environments.
+export function wrapInClientProxy({id, name, named, component}: ClientProxy) {
+  if (!isReactComponent(component, name)) {
```

Comment:
It's an important feature to be able to pass any reference to the client. It's not just React components. You can pass any value exported from a client module as a prop. This is important for dependency injection, passing mutable refs, passing references to large values that are available on the client anyway and aren't sent again and again.

Ideally this would also include "string" and "number" exports which can't be reliably wrapped in Proxies.